pandas/io/parsers.py

```diff
@@ -1146,18 +1146,27 @@ def __init__(self, src, **kwds):
             else:
                 self.names = lrange(self._reader.table_width)
 
-        # If the names were inferred (not passed by user) and usedcols is
-        # defined, then ensure names refers to the used columns, not the
-        # document's columns.
-        if self.usecols and passed_names:
-            col_indices = []
-            for u in self.usecols:
-                if isinstance(u, string_types):
-                    col_indices.append(self.names.index(u))
-                else:
-                    col_indices.append(u)
-            self.names = [n for i, n in enumerate(self.names)
-                          if i in col_indices]
+        # gh-9755
+        #
+        # need to set orig_names here first
+        # so that proper indexing can be done
+        # with _set_noconvert_columns
+        #
+        # once names has been filtered, we will
+        # then set orig_names again to names
+        self.orig_names = self.names[:]
+
+        if self.usecols:
+            if len(self.names) > len(self.usecols):
+                col_indices = []
+                for u in self.usecols:
+                    if isinstance(u, string_types):
```

Comment:
I decided to go with the one-liner I suggested earlier because your suggestion unfortunately is actually an API change.  In the original loop, the order of the columns in the original data is preserved _independent_ of the order of `usecols`, whereas `take(indexer)` makes it _dependent_.  This causes several test failures like this one <a href="https://github.com/pydata/pandas/blob/master/pandas/io/tests/test_parsers.py#L2243">here</a>.